File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/visualization/scatterplot/ScatterplotOpDesc.java
package edu.uci.ics.texera.workflow.operators.visualization.scatterplot;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.kjetland.jackson.jsonSchema.annotations.JsonSchemaInject;
import com.kjetland.jackson.jsonSchema.annotations.JsonSchemaTitle;
import edu.uci.ics.amber.engine.architecture.deploysemantics.PhysicalOp;
import edu.uci.ics.amber.engine.architecture.deploysemantics.layer.OpExecInitInfo;
import edu.uci.ics.amber.engine.architecture.scheduling.config.OperatorConfig;
import edu.uci.ics.amber.engine.common.IOperatorExecutor;
import edu.uci.ics.amber.engine.common.virtualidentity.ExecutionIdentity;
import edu.uci.ics.amber.engine.common.virtualidentity.WorkflowIdentity;
import edu.uci.ics.texera.workflow.common.metadata.OperatorGroupConstants;
import edu.uci.ics.texera.workflow.common.metadata.OperatorInfo;
import edu.uci.ics.texera.workflow.common.metadata.annotations.AutofillAttributeName;
import edu.uci.ics.texera.workflow.common.tuple.schema.Attribute;
import edu.uci.ics.texera.workflow.common.tuple.schema.AttributeType;
import edu.uci.ics.texera.workflow.common.tuple.schema.Schema;
import edu.uci.ics.amber.engine.common.workflow.InputPort;
import edu.uci.ics.amber.engine.common.workflow.OutputPort;
import edu.uci.ics.amber.engine.common.workflow.PortIdentity;
import edu.uci.ics.texera.workflow.operators.visualization.VisualizationConstants;
import edu.uci.ics.texera.workflow.operators.visualization.VisualizationOperator;
import scala.Tuple3;
import scala.collection.immutable.List;

import java.util.EnumSet;
import java.util.Set;
import java.util.function.Function;

import static edu.uci.ics.texera.workflow.common.tuple.schema.AttributeType.DOUBLE;
import static edu.uci.ics.texera.workflow.common.tuple.schema.AttributeType.INTEGER;
import static java.util.Collections.singletonList;
import static scala.collection.JavaConverters.asScalaBuffer;

/**
 * Scatterplot operator to visualize the result as a scatterplot
 * This is the description of the operator
 */

@JsonSchemaInject(json =
        "{" +
                "  \"attributeTypeRules\": {" +
                "    \"xColumn\":{" +
                "      \"enum\": [\"integer\", \"double\"]" +
                "    }," +
                "    \"yColumn\":{" +
                "      \"enum\": [\"integer\", \"double\"]" +
                "    }" +
                "  }" +
                "}")
public class ScatterplotOpDesc extends VisualizationOperator {
    @JsonProperty(required = true)
    @JsonSchemaTitle("X-Column")
    @AutofillAttributeName
    public String xColumn;

    @JsonProperty(required = true)
    @JsonSchemaTitle("Y-Column")
    @AutofillAttributeName
    public String yColumn;

    @JsonProperty(defaultValue = "false")
    @JsonSchemaTitle("Geometric")
    @JsonPropertyDescription("plot on a map")
    public boolean isGeometric;

    @Override
    public String chartType() {
        if (isGeometric) {
            return VisualizationConstants.SPATIAL_SCATTERPLOT;
        }
        return VisualizationConstants.SIMPLE_SCATTERPLOT;
    }

    @Override
    public PhysicalOp getPhysicalOp(WorkflowIdentity workflowId, ExecutionIdentity executionId) {
        Schema inputSchema = this.inputPortToSchemaMapping().get(this.operatorInfo().inputPorts().head().id()).get();
        AttributeType xType = inputSchema.getAttribute(xColumn).getType();
        AttributeType yType = inputSchema.getAttribute(yColumn).getType();
        Set<AttributeType> allowedAttributeTypesNumbersOnly = EnumSet.of(DOUBLE, INTEGER); //currently, the frontend has limitation it doesn't accept axes of type long
        if (!allowedAttributeTypesNumbersOnly.contains(xType)) {
            throw new IllegalArgumentException(xColumn + " is not a number \n");
        }
        if (!allowedAttributeTypesNumbersOnly.contains(yType)) {
            throw new IllegalArgumentException(yColumn + " is not a number \n");
        }

        return PhysicalOp.oneToOnePhysicalOp(
                    workflowId,
                    executionId,
                    this.operatorIdentifier(),
                    OpExecInitInfo.apply(
                            (Function<Tuple3<Object, PhysicalOp, OperatorConfig>, IOperatorExecutor> & java.io.Serializable)
                                    worker -> new ScatterplotOpExec(this)
                    )
                )
                .withInputPorts(operatorInfo().inputPorts(), inputPortToSchemaMapping())
                .withOutputPorts(operatorInfo().outputPorts(), outputPortToSchemaMapping())
                .withIsOneToManyOp(true)
                .withParallelizable(!isGeometric);
    }

    @Override
    public OperatorInfo operatorInfo() {
        return new OperatorInfo(
                "Scatterplot",
                "View the result in a scatterplot",
                OperatorGroupConstants.VISUALIZATION_GROUP(),
                asScalaBuffer(singletonList(new InputPort(new PortIdentity(0, false), "", false, List.empty()))).toList(),
                asScalaBuffer(singletonList(new OutputPort(new PortIdentity(0, false ), ""))).toList(),
                false,
                false,
                false,
                false
        );
    }

    @Override
    public Schema getOutputSchema(Schema[] schemas) {
        Schema inputSchema = schemas[0];
        if (isGeometric)
            return Schema.newBuilder().add(
                    new Attribute("xColumn", inputSchema.getAttribute(xColumn).getType()),
                    new Attribute("yColumn", inputSchema.getAttribute(yColumn).getType())
            ).build();
        else
            return Schema.newBuilder().add(
                    new Attribute(xColumn, inputSchema.getAttribute(xColumn).getType()),
                    new Attribute(yColumn, inputSchema.getAttribute(yColumn).getType())
            ).build();
    }
}

File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/visualization/scatterplot/ScatterplotOpDesc.scala
package edu.uci.ics.texera.workflow.operators.visualization.scatterplot

import com.fasterxml.jackson.annotation.{JsonProperty, JsonPropertyDescription}
import com.kjetland.jackson.jsonSchema.annotations.{JsonSchemaInject, JsonSchemaTitle}
import edu.uci.ics.amber.engine.common.workflow.{InputPort, OutputPort}
import edu.uci.ics.texera.workflow.common.metadata.annotations.AutofillAttributeName
import edu.uci.ics.texera.workflow.common.metadata.{OperatorGroupConstants, OperatorInfo}
import edu.uci.ics.texera.workflow.common.operators.PythonOperatorDescriptor
import edu.uci.ics.texera.workflow.common.tuple.schema.{Attribute, AttributeType, Schema}
import edu.uci.ics.texera.workflow.operators.visualization.{
  VisualizationConstants,
  VisualizationOperator
}

@JsonSchemaInject(
  json =
    "{" +
      "  \"attributeTypeRules\": {" +
      "    \"xColumn\":{" +
      "      \"enum\": [\"integer\", \"double\"]" +
      "    }," +
      "    \"yColumn\":{" +
      "      \"enum\": [\"integer\", \"double\"]" +
      "    }" +
      "  }" +
      "}"
)
class ScatterplotOpDesc extends VisualizationOperator with PythonOperatorDescriptor {

  @JsonProperty(required = true)
  @JsonSchemaTitle("X-Column")
  @JsonPropertyDescription("X Column")
  @AutofillAttributeName
  private val xColumn: String = ""

  @JsonProperty(required = true)
  @JsonSchemaTitle("Y-Column")
  @JsonPropertyDescription("Y Column")
  @AutofillAttributeName
  private val yColumn: String = ""

  @JsonProperty(required = false)
  @JsonSchemaTitle("Color-Column")
  @JsonPropertyDescription(
    "Dots will be assigned different colors based on their values of this column"
  )
  @AutofillAttributeName
  private val colorColumn: String = ""

  override def chartType: String = VisualizationConstants.HTML_VIZ

  override def getOutputSchema(schemas: Array[Schema]): Schema = {
    Schema.builder().add(new Attribute("html-content", AttributeType.STRING)).build()
  }

  override def operatorInfo: OperatorInfo =
    OperatorInfo(
      "Scatterplot",
      "View the result in a scatterplot",
      OperatorGroupConstants.VISUALIZATION_GROUP,
      inputPorts = List(InputPort()),
      outputPorts = List(OutputPort())
    )

  def manipulateTable(): String = {
    assert(xColumn.nonEmpty && yColumn.nonEmpty)
    val colorColExpr = if (colorColumn.nonEmpty) { s"'$colorColumn'" }
    else { "" }
    s"""
       |        # drops rows with missing values pertaining to relevant columns
       |        table.dropna(subset=['$xColumn', '$yColumn', $colorColExpr], inplace = True)
       |
       |""".stripMargin
  }

  def createPlotlyFigure(): String = {
    assert(xColumn.nonEmpty && yColumn.nonEmpty)
    val colorColExpr = if (colorColumn.nonEmpty) { s"color='$colorColumn'" }
    else { "" }
    s"""
           |        fig = go.Figure(px.scatter(table, x='$xColumn', y='$yColumn'))
           |        fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
           |""".stripMargin
  }

  override def generatePythonCode(): String = {
    val finalCode =
      s"""
           |from pytexera import *
           |
           |import plotly.express as px
           |import plotly.graph_objects as go
           |import plotly.io
           |import numpy as np
           |
           |
           |class ProcessTableOperator(UDFTableOperator):
           |
           |    def render_error(self, error_msg):
           |        return '''<h1>Scatter Plot is not available.</h1>
           |                  <p>Reasons are: {} </p>
           |               '''.format(error_msg)
           |
           |    @overrides
           |    def process_table(self, table: Table, port: int) -> Iterator[Optional[TableLike]]:
           |        if table.empty:
           |            yield {'html-content': self.render_error("Input table is empty.")}
           |            return
           |        ${manipulateTable()}
           |        ${createPlotlyFigure()}
           |        if table.empty:
           |            yield {'html-content': self.render_error("No valid rows left (every row has at least 1 missing value).")}
           |            return
           |        html = plotly.io.to_html(fig, include_plotlyjs = 'cdn', auto_play = False)
           |        yield {'html-content':html}
           |""".stripMargin
    finalCode
  }
}
